### PR TITLE
No need to send "Content-Disposition: inline" header

### PR DESCRIPTION
**Pull request by mkurz**
According to [RFC 6266 Section 4.2](https://tools.ietf.org/html/rfc6266#section-4.2) there is absolutely no need to send the header `Content-Disposition: inline`:

> On the other hand, if it matches "inline" (case-insensitively), this **implies default processing**. Therefore, **the disposition type "inline" is only useful when it is augmented with additional parameters**, such as the filename...

That should already be enough proof to skip this header, but here is a reference to the MDN web docs as well, [which also clearly state](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition#As_a_response_header_for_the_main_body) that `inline` is the default:

> The first parameter in the HTTP context is either inline (**default value**, indicating it can be displayed inside the Web page, or as the Web page) or attachment...

Also Drupal stopped sending this exact header a couple of years ago already: 
https://www.drupal.org/project/private_download/issues/2010092

However, because I wanted to be 100% sure that nothing changes if not sending that specific header anymore I checked if any given browser engine would act differently when sending/not sending this header: Fortunately it turned out that all browsers adhere to the RFC meaning it just does not matter if you send this header or not, every browser just cares about `Content-Disposition: attachment...` and some about `Content-Disposition: inline; filename=...`, but no browser cares solely about `Content-Disposition: inline`. So IMHO it's safe to just avoid sending this _exact_ header.

Here are the relevant lines in the browsers' source codes:

#### Gecko (Firefox)

Just cares about `attachment`:  [Lines 401-411](https://hg.mozilla.org/mozilla-central/file/d55401632cea92b6b2775ba278274b5490275876/uriloader/base/nsURILoader.cpp#l409), same [here](https://hg.mozilla.org/mozilla-central/file/73c98da145a7c0ef518404493b23a979f328768e/uriloader/exthandler/nsExternalHelperAppService.cpp#l215).

#### Chromium (Chrome and upcoming MS Edge)

If you scroll through [this file](https://chromium.googlesource.com/chromium.git/+/53fad3a9e293951ef1e54dccbe8a9cdea0d5fc8d/net/http/http_content_disposition.cc#65) you will see the default is set to `inline` and only explicitly sets "attachment" (defined [here](https://chromium.googlesource.com/chromium.git/+/801f316d276b9bdab0f8109aa1055cc2d2153e1e/net/http/http_content_disposition.h#17)). Also [here](https://chromium.googlesource.com/chromium.git/+/f2cf4e9b5ee7d1270fa9410da9a7a795a533292b/content/browser/renderer_host/buffered_resource_handler.cc#404) Chromium just cares about attachment.

#### Webkit (Safari)

Just cares about [isAttachment](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/platform/network/ResourceResponseBase.cpp?rev=239749#L746) and [isAttachmentWithFilename](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/platform/network/ResourceResponseBase.cpp?rev=239749#L754). "inline" just doesn't matter.

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @mergify[bot],

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by marcospereira** [comment]
> I see where @marcospereira is coming from, but I'm not sure if it's worth complicating the API

I think it is about simplifying it. Keeping `inline` when it is not useful besides `inline = false` which means `attachment` is less intuitive to me than have inlining as the default (not specified since it is unnecessary) and having `sendAttachment` where you know attachment will be used.

What do you think?

**Comment by mkurz** [comment]
FYI: I am thinking about this. I think there is another way how to solve this. I will share when I figured out more concrete details.

**Comment by mkurz** [comment]
I understand your intentions @marcospereira. I think the boolean parameter should have been named `attachment` in the first place (with a default of `false`) instead of `inline`, which would make this case here a bit easier. If it had been called `attachment`, it would be much clearer that someone wants the file to be an attachment or not. The naming `inline` doesn't fit here so well, because if I pass the method `inline = true`, someone would expect the inline will be send, but now Play won't because it somehow automagically  says that's not necessary because of specs. Having the param named `attachment` it wouldn't matter so much how to handle inline, because it wouldn't be emphasized very much (actually not at all). However renaming the boolean is not an option obviously, so this ship has sailed - or has it? We of course could change the boolean's position in the parameter list - or there is another way which I present you later in this comment. Read on... :wink: 
 
Your suggestion of adding `sendAttachment` also does not fit very well IMHO. I guess you want this method replace all the `sendFile`, `sendPath`,... methods (so we would have various overrides like `sendAttachment(File,...`, `sendAttachment(Path,...)`, etc.) But what If I don't want to send a file as attachment but as inline with a filename, ending up sending `Content-Disposition: inline; filename=...` (specs explicitly mentions this and browser do support this)? How would you do that? Clearly not with the `sendAttachment` method...

That's why I suggest to change the API just "slightly"... The original
```java
public Result sendPath(Path path, boolean inline, String filename, FileMimeTypes fileMimeTypes)
```
would become
```java
public Result sendPath(Path path, ContentDisposition mode, Optional<String> filename, FileMimeTypes fileMimeTypes) {
```
where `mode` would be an `enum` with values `ATTACHMENT`, `INLINE_LAX` and `INLINE_FORCE` (or similar, suggestions welcome). This enum would give users full control if the inline should be send even if it's not needed (because, you know, maybe some weird client out there needs this header set no matter what for some goddamn reason...). Default would be the specs friendly `INLINE_LAX` (or however we call it) which, like in my pull request, detects if the header really needs to be send and avoids it if there is no need for it.
Wrapping the `filename` into an `Optional` does make sense now in the light of recently merged #9349 (much nicer than passing `null` IMHO),

But wait! Now, if we would change `filename` to be an `Optional`, we could also go ahead and in the same step also just rename the content-disposition boolean from `inline` to `attachment`, so we move the emphasis away from inline, which possibly gives us more freedom how to interpret `attachment=false`... by, again, automagically not sending the header if there is not need for it. Consequences: No enum necessary, but also less user control, because Play magic decides if there is a need to send the header or not.
You get my idea?

To sum it up, my second suggestion to change the API would end up being:
```java
public Result sendPath(Path path, boolean attachment, Optional<String> filename, FileMimeTypes fileMimeTypes) {
```

Of course in the deprecation notes of the original methods we would have to tell users to not just wrap the filename into an `Optional` now, but tell them loud and clear that inline is attachment now, so probably they need to switch from `true` to `false` and vice versa.

The same I described here applies to the Scala API as well of course.

So @marcospereira what do you think? What would you prefer?
Your choice :smile:

**Comment by mkurz** [comment]
@marcospereira Thoughts? I would like to get a solution into the next milestone release...
(I somehow prefer the second option, without the enum, because someone can still use `.withHeader` anyway, and it's less verbose maybe)

**Comment by mkurz** [comment]
Hey @marcospereira! Looking forward to your feedback - I would like to get a solution for this into the next milestone release. Thanks a lot!

File: core/play/src/main/java/play/mvc/StatusHeader.java [hunk 1]
```diff
@@ -558,16 +558,20 @@ private Result doSendResource(
...
-          HttpHeaderParameterEncoding.encodeToBuilder("filename", rn, cdBuilder);
-        });
-    Map<String, String> headers =
-        Collections.singletonMap(Http.HeaderNames.CONTENT_DISPOSITION, cdBuilder.toString());
+    Map<String, String> headers = Collections.emptyMap();
+    if (!inline || resourceName.filter(rn -> !rn.isEmpty()).isPresent()) {
+      // Create a Content-Disposition header
+      // According to RFC 6266 (Section 4.2) there is no need to send "Content-Disposition: inline"
+      StringBuilder cdBuilder = new StringBuilder();
+      cdBuilder.append(inline ? "inline" : "attachment");
```
[the remark this answers is not in the document]
Comment (mkurz) [reply core/play/src/main/java/play/mvc/StatusHeader.java]: Yes, `inline` will be send when there is a `resourceName` and `inline = true`. There are tests in place for all of this.

File: core/play/src/main/java/play/mvc/StatusHeader.java [hunk 2]
```diff
@@ -558,16 +558,20 @@ private Result doSendResource(
...
-        rn -> {
-          cdBuilder.append("; ");
-          HttpHeaderParameterEncoding.encodeToBuilder("filename", rn, cdBuilder);
-        });
-    Map<String, String> headers =
-        Collections.singletonMap(Http.HeaderNames.CONTENT_DISPOSITION, cdBuilder.toString());
+    Map<String, String> headers = Collections.emptyMap();
+    if (!inline || resourceName.filter(rn -> !rn.isEmpty()).isPresent()) {
+      // Create a Content-Disposition header
+      // According to RFC 6266 (Section 4.2) there is no need to send "Content-Disposition: inline"
```
[the remark this answers is not in the document]
Comment (mkurz) [reply core/play/src/main/java/play/mvc/StatusHeader.java]: Added the link.

**Comment by mkurz** [comment]
Alright, after I had a look again I think we should just merge this pull request and its done :wink: 
I now added migration notes too, even though migration steps wont be necessary by users really, but to make them aware of this "change".

I don't think its worth to add an `enum` like mentioned in my previous comment. It's just too much hassle and doesn't really add any value (because you can still add the header via `withHeader` anyway).
My other suggestion of changing `boolean inline` to `boolean attachment` doesn't work for all methods, because [we have method signatures](https://github.com/playframework/playframework/blob/8b0d5afb8c353dd8cd8d9e8057136e1858ad0173/core/play/src/main/java/play/mvc/StatusHeader.java#L463) which wouldn't allow this change (user would have to change `true` to `false` and vice versa to migrate, however no warning would be raised, therefore not backwards compatible).
Also switching to Marco's suggested `sendAttachment` method wouldn't work, because we still would need a way to send files inline.

I do agree with Marco's statement:
> ...`inline` when it is not useful besides `inline = false` which means `attachment` is less intuitive to me...

The `boolean` should have been named `attachment` in the first place, which would be the most intuitive way to handle all of this, but without major API change, this ship has sailed.
